<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Annie Xie" />

<meta name="date" content="2025-09-11" />

<title>bal_nonoverlap_setting</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">stability_selection</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/xxie6/stability_selection">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">bal_nonoverlap_setting</h1>
<h4 class="author">Annie Xie</h4>
<h4 class="date">2025-09-11</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-09-17
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>stability_selection/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20250906code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20250906)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20250906code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20250906)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomxxie6stabilityselectiontree8549102e9576ec0d916c4e191a6fdcf750abaf8etargetblank8549102a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/xxie6/stability_selection/tree/8549102e9576ec0d916c4e191a6fdcf750abaf8e" target="_blank">8549102</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomxxie6stabilityselectiontree8549102e9576ec0d916c4e191a6fdcf750abaf8etargetblank8549102a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/xxie6/stability_selection/tree/8549102e9576ec0d916c4e191a6fdcf750abaf8e" target="_blank">8549102</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Unstaged changes:
    Modified:   analysis/baltree_setting.Rmd
    Modified:   analysis/sparse_overlap_setting.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/bal_nonoverlap_setting.Rmd</code>) and HTML
(<code>docs/bal_nonoverlap_setting.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/xxie6/stability_selection/blob/8549102e9576ec0d916c4e191a6fdcf750abaf8e/analysis/bal_nonoverlap_setting.Rmd" target="_blank">8549102</a>
</td>
<td>
Annie Xie
</td>
<td>
2025-09-17
</td>
<td>
Update row-split analysis in bal nonoverlap setting
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/xxie6/stability_selection/3532ad3890e7c167ffbaacc92157bcf5f97980a2/docs/bal_nonoverlap_setting.html" target="_blank">3532ad3</a>
</td>
<td>
Annie Xie
</td>
<td>
2025-09-11
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/xxie6/stability_selection/blob/9962de93da83f849c64b6ef51e66a2a47cefca6f/analysis/bal_nonoverlap_setting.Rmd" target="_blank">9962de9</a>
</td>
<td>
Annie Xie
</td>
<td>
2025-09-11
</td>
<td>
Add exporation of balanced non-overlap setting
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>In this analysis, we are interested in testing stability selection
approaches in the balanced non-overlapping setting. At a high level, the
stability selection involves 1) splitting the data into two subsets, 2)
applying the method to each subset, 3) choosing the components that have
high correspondence across the two sets of results. We will test two
different approaches to step 1). The first approach is splitting the
data by splitting the columns. This approach feels intuitive since we
are interested in the loadings matrix, which says something about the
samples in the dataset. In a population genetics application, one could
argue that all of the chromosomes are undergoing evolution
independently, and so you could split the data by even vs. odd
chromosomes to get two different datasets. However, in a single-cell
RNA-seq application, it feels more natural to split the data by cells –
this feels more like creating sub-datasets compared to splitting by
genes (unless you want to make some assumption that the genes are pulled
from a “population”, but I think that feels less natural). This
motivates the second approach: splitting the data by splitting the
rows.</p>
<pre class="r"><code>library(dplyr)
library(ggplot2)
library(pheatmap)</code></pre>
<pre class="r"><code>source(&#39;code/visualization_functions.R&#39;)
source(&#39;code/stability_selection_functions.R&#39;)</code></pre>
</div>
<div id="data-generation" class="section level1">
<h1>Data Generation</h1>
<p>In this analysis, we will focus on the balanced non-overlapping
setting.</p>
<pre class="r"><code>sim_star_data &lt;- function(args) {
  set.seed(args$seed)
  
  n &lt;- sum(args$pop_sizes)
  p &lt;- args$n_genes
  K &lt;- length(args$pop_sizes)
  
  FF &lt;- matrix(rnorm(K * p, sd = rep(args$branch_sds, each = p)), ncol = K)
  
  LL &lt;- matrix(0, nrow = n, ncol = K)
  for (k in 1:K) {
    vec &lt;- rep(0, K)
    vec[k] &lt;- 1
    LL[, k] &lt;- rep(vec, times = args$pop_sizes)
  }
  
  E &lt;- matrix(rnorm(n * p, sd = args$indiv_sd), nrow = n)
  Y &lt;- LL %*% t(FF) + E
  YYt &lt;- (1/p)*tcrossprod(Y)
  
  return(list(Y = Y, YYt = YYt, LL = LL, FF = FF, K = ncol(LL)))
}</code></pre>
<pre class="r"><code>pop_sizes &lt;- rep(40,4)
n_genes &lt;- 1000
branch_sds &lt;- rep(2,4)
indiv_sd &lt;- 1
seed &lt;- 1
sim_args = list(pop_sizes = pop_sizes, branch_sds = branch_sds, indiv_sd = indiv_sd, n_genes = n_genes, seed = seed)
sim_data &lt;- sim_star_data(sim_args)</code></pre>
<p>This is a heatmap of the true loadings matrix:</p>
<pre class="r"><code>plot_heatmap(sim_data$LL)</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-5-1">
Past versions of unnamed-chunk-5-1.png
</button>
</p>
<div id="fig-unnamed-chunk-5-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/stability_selection/blob/3532ad3890e7c167ffbaacc92157bcf5f97980a2/docs/figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-5-1.png" target="_blank">3532ad3</a>
</td>
<td>
Annie Xie
</td>
<td>
2025-09-11
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="gbcd" class="section level1">
<h1>GBCD</h1>
<p>In this section, I try stability selection with the GBCD method. In
my experiments, I’ve found that GBCD tends to return extra factors
(partially because the point-Laplace initialization will yield extra
factors).</p>
<div id="stability-selection-via-splitting-columns"
class="section level2">
<h2>Stability Selection via Splitting Columns</h2>
<p>First, I try splitting the data by splitting the columns.</p>
<pre class="r"><code>set.seed(1)
X_split_by_col &lt;- stability_selection_split_data(sim_data$Y, dim = &#39;columns&#39;)</code></pre>
<pre class="r"><code>gbcd_fits_by_col &lt;- list()
for (i in 1:(length(X_split_by_col)-2)){
  gbcd_fits_by_col[[i]] &lt;- gbcd::fit_gbcd(X_split_by_col[[i]], 
                                   Kmax = 4, 
                                   prior = ebnm::ebnm_generalized_binary,
                                   verbose = 0)$L
}</code></pre>
<pre><code>[1] &quot;Form cell by cell covariance matrix...&quot;
   user  system elapsed 
  0.005   0.001   0.005 
[1] &quot;Initialize GEP membership matrix L...&quot;
Adding factor 1 to flash object...
Wrapping up...
Done.
Adding factor 2 to flash object...
Adding factor 3 to flash object...
Adding factor 4 to flash object...
Wrapping up...
Done.</code></pre>
<pre><code>Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.</code></pre>
<pre><code>   user  system elapsed 
  4.936   0.116   5.055 
[1] &quot;Estimate GEP membership matrix L...&quot;</code></pre>
<pre><code>Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.
Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.
Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.</code></pre>
<pre><code>   user  system elapsed 
 15.142   0.185  15.354 
[1] &quot;Estimate GEP signature matrix F...&quot;
Backfitting 6 factors (tolerance: 1.19e-03)...
  Difference between iterations is within 1.0e+00...
An update to factor 2 decreased the objective by 2.910e-11.
An update to factor 2 decreased the objective by -0.000e+00.
An update to factor 2 decreased the objective by 2.910e-11.
An update to factor 2 decreased the objective by -0.000e+00.
An update to factor 2 decreased the objective by 8.731e-11.
An update to factor 2 decreased the objective by 2.910e-11.
An update to factor 2 decreased the objective by -0.000e+00.
An update to factor 2 decreased the objective by 2.910e-11.
An update to factor 2 decreased the objective by 2.910e-11.
An update to factor 2 decreased the objective by -0.000e+00.
An update to factor 2 decreased the objective by 1.455e-11.
  --Maximum number of iterations reached!
Wrapping up...
Done.
   user  system elapsed 
  8.767   0.235   9.003 
[1] &quot;Form cell by cell covariance matrix...&quot;
   user  system elapsed 
  0.003   0.000   0.003 
[1] &quot;Initialize GEP membership matrix L...&quot;
Adding factor 1 to flash object...
Wrapping up...
Done.
Adding factor 2 to flash object...
Adding factor 3 to flash object...
Adding factor 4 to flash object...
Wrapping up...
Done.</code></pre>
<pre><code>Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.</code></pre>
<pre><code>   user  system elapsed 
  2.541   0.069   2.611 
[1] &quot;Estimate GEP membership matrix L...&quot;</code></pre>
<pre><code>Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.
Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.
Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.</code></pre>
<pre><code>   user  system elapsed 
 14.559   0.163  14.722 
[1] &quot;Estimate GEP signature matrix F...&quot;
Backfitting 6 factors (tolerance: 1.19e-03)...
  Difference between iterations is within 1.0e+01...
  --Estimate of factor 6 is numerically zero!
  Difference between iterations is within 1.0e+00...
  Difference between iterations is within 1.0e-01...
  --Estimate of factor 5 is numerically zero!
  Difference between iterations is within 1.0e-02...
Wrapping up...
Done.
   user  system elapsed 
  5.873   0.151   6.025 </code></pre>
<p>This is heatmap of the loadings estimate from the first subset:</p>
<pre class="r"><code>plot_heatmap(gbcd_fits_by_col[[1]], colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(gbcd_fits_by_col[[1]])), max(abs(gbcd_fits_by_col[[1]])), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-8-1">
Past versions of unnamed-chunk-8-1.png
</button>
</p>
<div id="fig-unnamed-chunk-8-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/stability_selection/blob/3532ad3890e7c167ffbaacc92157bcf5f97980a2/docs/figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-8-1.png" target="_blank">3532ad3</a>
</td>
<td>
Annie Xie
</td>
<td>
2025-09-11
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This is a heatmap of the loadings estimate from the second
subset:</p>
<pre class="r"><code>plot_heatmap(gbcd_fits_by_col[[2]], colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(gbcd_fits_by_col[[2]])), max(abs(gbcd_fits_by_col[[2]])), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-9-1">
Past versions of unnamed-chunk-9-1.png
</button>
</p>
<div id="fig-unnamed-chunk-9-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/stability_selection/blob/3532ad3890e7c167ffbaacc92157bcf5f97980a2/docs/figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-9-1.png" target="_blank">3532ad3</a>
</td>
<td>
Annie Xie
</td>
<td>
2025-09-11
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>One observation is GBCD returns duplicate factors. Why?</p>
<pre class="r"><code>results_by_col &lt;- stability_selection_post_processing(gbcd_fits_by_col[[1]], gbcd_fits_by_col[[2]], threshold = 0.8)
L_est_by_col &lt;- results_by_col$L</code></pre>
<p>This is the similarity matrix:</p>
<pre class="r"><code>results_by_col$similarity</code></pre>
<pre><code>             Baseline         GEP1         GEP2         GEP3         GEP4
Baseline  0.002036051 0.0008702023 6.789660e-27 9.986165e-01 9.987837e-01
GEP1      0.999018321 0.0035627650 2.911801e-07 1.470870e-03 4.704740e-03
GEP2     -0.008832349 0.9711621318 1.299493e-01 2.290885e-02 2.482999e-02
GEP3      0.007283325 0.0090558806 9.980530e-01 3.511246e-12 1.734330e-11
GEP4      0.002992929 0.9332134390 1.461648e-24 7.347701e-02 6.958626e-02
GEP5      0.007391862 0.0090400999 9.985535e-01 3.448226e-12 1.703617e-11
                GEP5
Baseline  0.23560380
GEP1      0.09232634
GEP2      0.06501785
GEP3      0.94736512
GEP4     -0.03882397
GEP5      0.94763481</code></pre>
<p>This is a heatmap of the final loadings estimate:</p>
<pre class="r"><code>plot_heatmap(L_est_by_col, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L_est_by_col)), max(abs(L_est_by_col)), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This is the correlation between the estimate and the true loadings
matrix:</p>
<pre class="r"><code>apply(cor(L_est_by_col, sim_data$LL), 1, max)</code></pre>
<pre><code> Baseline      GEP1      GEP2      GEP3      GEP4      GEP5 
0.9997589 0.9994970 0.9616086 0.9994997 0.9119285 0.9999360 </code></pre>
<p>In this case, the method was able to get rid of the extra factors and
recover all four components. However, one issue is a couple of the
components are repeated, meaning we have redundant factors. I could
enforce a 1-1 mapping in the stability selection. Though that will beg
the question of how a 1-1 mapping is chosen. Depending on the criterion,
I don’t think it’s guaranteed that two factors that have a similarity
greater than 0.99 will be paired (I think I have an example where this
is the case). Perhaps we need another procedure to remove redundant
factors? I can also look into why GBCD returns redundant factors since
this behavior is unexpected.</p>
</div>
<div id="stability-selection-via-splitting-rows" class="section level2">
<h2>Stability Selection via Splitting Rows</h2>
<p>Now, we try splitting the rows:</p>
<pre class="r"><code>set.seed(1)
X_split_by_row &lt;- stability_selection_split_data(sim_data$Y, dim = &#39;rows&#39;)</code></pre>
<pre class="r"><code>gbcd_fits_by_row_F &lt;- list()
gbcd_fits_by_row_L &lt;- list()
for (i in 1:(length(X_split_by_row)-2)){
  gbcd_fit &lt;- gbcd::fit_gbcd(X_split_by_row[[i]], 
                                   Kmax = 4, 
                                   prior = ebnm::ebnm_generalized_binary,
                                   verbose = 0)
  gbcd_fits_by_row_F[[i]] &lt;- gbcd_fit$F$lfc
  gbcd_fits_by_row_L[[i]] &lt;- gbcd_fit$L
}</code></pre>
<pre><code>[1] &quot;Form cell by cell covariance matrix...&quot;
   user  system elapsed 
  0.002   0.000   0.002 
[1] &quot;Initialize GEP membership matrix L...&quot;
Adding factor 1 to flash object...
Wrapping up...
Done.
Adding factor 2 to flash object...
Adding factor 3 to flash object...
Adding factor 4 to flash object...
Wrapping up...
Done.</code></pre>
<pre><code>Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.</code></pre>
<pre><code>   user  system elapsed 
  1.530   0.060   1.591 
[1] &quot;Estimate GEP membership matrix L...&quot;
   user  system elapsed 
  2.786   0.040   2.827 
[1] &quot;Estimate GEP signature matrix F...&quot;
Backfitting 5 factors (tolerance: 1.19e-03)...
  --Estimate of factor 5 is numerically zero!
  Difference between iterations is within 1.0e+01...
  Difference between iterations is within 1.0e+00...
Wrapping up...
Done.
   user  system elapsed 
  0.696   0.007   0.703 
[1] &quot;Form cell by cell covariance matrix...&quot;
   user  system elapsed 
  0.002   0.000   0.002 
[1] &quot;Initialize GEP membership matrix L...&quot;
Adding factor 1 to flash object...
Wrapping up...
Done.
Adding factor 2 to flash object...
Adding factor 3 to flash object...
Adding factor 4 to flash object...
Wrapping up...
Done.</code></pre>
<pre><code>Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.</code></pre>
<pre><code>   user  system elapsed 
  1.100   0.036   1.136 
[1] &quot;Estimate GEP membership matrix L...&quot;</code></pre>
<pre><code>Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.</code></pre>
<pre><code>   user  system elapsed 
  2.303   0.032   2.334 
[1] &quot;Estimate GEP signature matrix F...&quot;
Backfitting 7 factors (tolerance: 1.19e-03)...
  --Estimate of factor 7 is numerically zero!
  --Estimate of factor 5 is numerically zero!
  --Estimate of factor 6 is numerically zero!
  Difference between iterations is within 1.0e+01...
  Difference between iterations is within 1.0e+00...
Wrapping up...
Done.
   user  system elapsed 
  1.004   0.019   1.025 </code></pre>
<p>This is heatmap of the factor estimate from the first subset:</p>
<pre class="r"><code>plot_heatmap(gbcd_fits_by_row_L[[1]][order(X_split_by_row[[3]]),], colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(gbcd_fits_by_row_L[[1]])), max(abs(gbcd_fits_by_row_L[[1]])), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-16-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-16-1">
Past versions of unnamed-chunk-16-1.png
</button>
</p>
<div id="fig-unnamed-chunk-16-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/stability_selection/blob/3532ad3890e7c167ffbaacc92157bcf5f97980a2/docs/figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-16-1.png" target="_blank">3532ad3</a>
</td>
<td>
Annie Xie
</td>
<td>
2025-09-11
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This is heatmap of the factor estimate from the second subset:</p>
<pre class="r"><code>plot_heatmap(gbcd_fits_by_row_L[[2]], colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(gbcd_fits_by_row_L[[2]])), max(abs(gbcd_fits_by_row_L[[2]])), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-17-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-17-1">
Past versions of unnamed-chunk-17-1.png
</button>
</p>
<div id="fig-unnamed-chunk-17-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/stability_selection/blob/3532ad3890e7c167ffbaacc92157bcf5f97980a2/docs/figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-17-1.png" target="_blank">3532ad3</a>
</td>
<td>
Annie Xie
</td>
<td>
2025-09-11
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>results_by_row &lt;- stability_selection_row_split_post_processing(gbcd_fits_by_row_L[[1]], gbcd_fits_by_row_L[[2]], gbcd_fits_by_row_F[[1]], gbcd_fits_by_row_F[[2]], threshold = 0.8)
L_est_by_row &lt;- results_by_row$L
L_est_by_row &lt;- L_est_by_row[order(c(X_split_by_row[[3]], X_split_by_row[[4]])), ]</code></pre>
<p>This is the similarity matrix:</p>
<pre class="r"><code>results_by_row$similarity</code></pre>
<pre><code>            Baseline        GEP1       GEP2       GEP3 GEP4 GEP5 GEP6
Baseline 0.024786123 0.011972328 0.05024929 0.98897840    0    0    0
GEP1     0.007202973 0.017581725 0.98847905 0.03025033    0    0    0
GEP2     0.017166495 0.989082679 0.02201011 0.01645547    0    0    0
GEP3     0.986741086 0.007415112 0.02657000 0.01656971    0    0    0
GEP4     0.000000000 0.000000000 0.00000000 0.00000000    0    0    0</code></pre>
<p>This is a heatmap of the final loadings estimate:</p>
<pre class="r"><code>plot_heatmap(L_est_by_row, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L_est_by_row)), max(abs(L_est_by_row)), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This is the correlation between the estimate and true loadings
matrix:</p>
<pre class="r"><code>apply(cor(L_est_by_row, sim_data$LL), 1, max)</code></pre>
<pre><code> Baseline      GEP1      GEP2      GEP3 
0.9995999 0.9995610 0.9995389 0.9994453 </code></pre>
<p>The method recovers all four components with no redundant
factors.</p>
</div>
</div>
<div id="ebcd" class="section level1">
<h1>EBCD</h1>
<p>In this section, I try stability selection with the GBCD method. When
given a <code>Kmax</code> value that is larger than the true number of
components, I’ve found that EBCD usually returns extra factors. So in
this section, when I run EBCD, I give the method double the true number
of components.</p>
<div id="stability-selection-via-splitting-columns-1"
class="section level2">
<h2>Stability Selection via Splitting Columns</h2>
<pre class="r"><code>set.seed(1)
ebcd_fits_by_col &lt;- list()
for (i in 1:(length(X_split_by_col)-2)){
  ebcd_fits_by_col[[i]] &lt;- ebcd::ebcd(t(X_split_by_col[[i]]), 
                                   Kmax = 8, 
                                   ebnm_fn = ebnm::ebnm_generalized_binary)$EL
}</code></pre>
<p>This is a heatmap of the estimated loadings from the first
subset:</p>
<pre class="r"><code>plot_heatmap(ebcd_fits_by_col[[1]], colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(ebcd_fits_by_col[[1]])), max(abs(ebcd_fits_by_col[[1]])), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-23-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-23-1">
Past versions of unnamed-chunk-23-1.png
</button>
</p>
<div id="fig-unnamed-chunk-23-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/stability_selection/blob/3532ad3890e7c167ffbaacc92157bcf5f97980a2/docs/figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-23-1.png" target="_blank">3532ad3</a>
</td>
<td>
Annie Xie
</td>
<td>
2025-09-11
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This is a heatmap of the estimated loadings from the second
subset:</p>
<pre class="r"><code>plot_heatmap(ebcd_fits_by_col[[2]], colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(ebcd_fits_by_col[[2]])), max(abs(ebcd_fits_by_col[[2]])), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-24-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-24-1">
Past versions of unnamed-chunk-24-1.png
</button>
</p>
<div id="fig-unnamed-chunk-24-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/stability_selection/blob/3532ad3890e7c167ffbaacc92157bcf5f97980a2/docs/figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-24-1.png" target="_blank">3532ad3</a>
</td>
<td>
Annie Xie
</td>
<td>
2025-09-11
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>results_by_col &lt;- stability_selection_post_processing(ebcd_fits_by_col[[1]], ebcd_fits_by_col[[2]], threshold = 0.8)
L_est_by_col &lt;- results_by_col$L</code></pre>
<p>This is the similarity matrix:</p>
<pre class="r"><code>results_by_col$similarity</code></pre>
<pre><code>             l            l            l            l            l            l
l 1.291968e-16 1.739759e-20 9.993891e-01 4.920096e-21 1.480176e-01 9.811116e-02
l 1.711919e-16 9.997425e-01 9.121040e-21 3.167954e-20 2.238270e-01 2.295285e-01
l 2.697388e-20 3.986695e-19 1.227701e-20 9.995145e-01 1.504057e-01 9.473680e-02
l 9.995149e-01 2.853284e-17 2.200194e-17 9.148350e-21 1.472496e-01 1.300156e-01
l 1.534390e-01 1.136690e-01 2.384712e-01 1.479971e-01 5.578555e-02 6.603956e-02
l 1.659964e-01 1.758761e-01 1.678360e-01 8.246162e-02 7.520223e-02 8.238879e-02
l 1.327326e-01 1.360105e-01 1.373094e-01 1.412127e-01 1.504386e-06 1.690897e-01
l 1.568373e-01 1.168567e-01 1.985823e-01 1.557710e-01 1.762884e-01 2.298739e-07
             l          l
l 0.1001376113 0.16737937
l 0.0565540906 0.20915092
l 0.1896297261 0.08494216
l 0.1536085481 0.12784075
l 0.0668537915 0.06101997
l 0.0771582627 0.27704113
l 0.0776890020 0.07512826
l 0.0001413651 0.14222217</code></pre>
<p>This is a heatmap of the final loadings estimate:</p>
<pre class="r"><code>plot_heatmap(L_est_by_col, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L_est_by_col)), max(abs(L_est_by_col)), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-27-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This is the correlation between the estimate and the true loadings
matrix:</p>
<pre class="r"><code>apply(cor(L_est_by_col, sim_data$LL), 1, max)</code></pre>
<pre><code>        l         l         l         l 
0.9996634 0.9997954 0.9997039 0.9997178 </code></pre>
<p>The method recovers all four components.</p>
</div>
<div id="stability-selection-via-splitting-rows-1"
class="section level2">
<h2>Stability Selection via Splitting Rows</h2>
<pre class="r"><code>transform_ebcd_Z &lt;- function(Y, ebcd_obj){
  Y.svd &lt;- svd(Y)
  Y.UV &lt;- Y.svd$u %*% t(Y.svd$v)
  Z_transformed &lt;- Y.UV %*% ebcd_obj$Z
  return(Z_transformed)
}</code></pre>
<pre class="r"><code>set.seed(1)
ebcd_fits_by_row_L &lt;- list()
ebcd_fits_by_row_F &lt;- list()
for (i in 1:(length(X_split_by_row)-2)){
  ebcd_fit &lt;- ebcd::ebcd(t(X_split_by_row[[i]]), 
                                   Kmax = 8, 
                                   ebnm_fn = ebnm::ebnm_generalized_binary)
  ebcd_fits_by_row_L[[i]] &lt;- ebcd_fit$EL
  ebcd_fits_by_row_F[[i]] &lt;- transform_ebcd_Z(t(X_split_by_row[[i]]), ebcd_fit)
}</code></pre>
<p>This is a heatmap of the loadings estimate from the first subset:</p>
<pre class="r"><code>plot_heatmap(ebcd_fits_by_row_L[[1]][order(X_split_by_row[[3]]),], colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(ebcd_fits_by_row_L[[1]])), max(abs(ebcd_fits_by_row_L[[1]])), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-31-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-31-1">
Past versions of unnamed-chunk-31-1.png
</button>
</p>
<div id="fig-unnamed-chunk-31-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/stability_selection/blob/3532ad3890e7c167ffbaacc92157bcf5f97980a2/docs/figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-31-1.png" target="_blank">3532ad3</a>
</td>
<td>
Annie Xie
</td>
<td>
2025-09-11
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This is a heatmap of the loadings estimate from the second
subset:</p>
<pre class="r"><code>plot_heatmap(ebcd_fits_by_row_L[[2]], colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(ebcd_fits_by_row_L[[2]])), max(abs(ebcd_fits_by_row_L[[2]])), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-32-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>results_by_row &lt;- stability_selection_row_split_post_processing(ebcd_fits_by_row_L[[1]], ebcd_fits_by_row_L[[2]], ebcd_fits_by_row_F[[1]], ebcd_fits_by_row_F[[2]], threshold = 0.5)
L_est_by_row &lt;- results_by_row$L
L_est_by_row &lt;- L_est_by_row[order(c(X_split_by_row[[3]], X_split_by_row[[4]])), ]</code></pre>
<p>This is the similarity matrix:</p>
<pre class="r"><code>results_by_row$similarity</code></pre>
<pre><code>              [,1]         [,2]         [,3]         [,4]          [,5]
[1,]  0.0032102127  0.967802286 -0.005102233  0.007089620 -0.0003627249
[2,]  0.0027556207  0.003719887  0.987639364 -0.004588882  0.0170573129
[3,] -0.0001969255 -0.010140331  0.001466850  0.986046031  0.0478663365
[4,]  0.9833435354 -0.009771908 -0.005755508 -0.005569431  0.0491616191
[5,]  0.0353328483  0.037055783  0.029676615  0.028481535 -0.0067407975
[6,]  0.0207562518  0.010759857  0.030412915  0.019966878  0.0020821776
[7,]  0.0254423042  0.043288068  0.006259009  0.019857270  0.0555537972
[8,]  0.0336043415  0.018256737  0.025816087  0.037233769 -0.0133825512
              [,6]          [,7]         [,8]
[1,]  0.1903913249  0.0004155998  0.034075536
[2,] -0.0044054072  0.0010000253  0.001955857
[3,] -0.0018373071  0.0030180076  0.003436489
[4,]  0.0191720195  0.0217717180 -0.003160007
[5,]  0.0238258276 -0.0099977254  0.059545492
[6,]  0.0004457379 -0.0353981535 -0.002271570
[7,]  0.0086231060  0.0143904386 -0.023355103
[8,] -0.0157438792  0.0250471478 -0.044566871</code></pre>
<p>This is a heatmap of the final loadings estimate:</p>
<pre class="r"><code>plot_heatmap(L_est_by_row, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L_est_by_row)), max(abs(L_est_by_row)), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-35-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This is the correlation between the estimate and the true
loadings:</p>
<pre class="r"><code>apply(cor(L_est_by_row, sim_data$LL), 1, max)</code></pre>
<pre><code>        l         l         l         l 
0.9997848 0.9998442 0.9998607 0.9998386 </code></pre>
<p>This method also recovered all four components.</p>
</div>
</div>
<div id="codesymnmf" class="section level1">
<h1>CoDesymNMF</h1>
<p>In this section, I try stability selection with the CoDesymNMF
method. Similar to EBCD, when given a <code>Kmax</code> value that is
larger than the true number of components, the method usually returns
extra factors. Note that in this section, when I run CoDesymNMF, I give
the method double the true number of components.</p>
<div id="stability-selection-via-splitting-columns-2"
class="section level2">
<h2>Stability Selection via Splitting Columns</h2>
<pre class="r"><code>codesymnmf_fits_by_col &lt;- list()
for (i in 1:(length(X_split_by_col)-2)){
  cov_mat &lt;- tcrossprod(X_split_by_col[[i]])/ncol(X_split_by_col[[i]])
  codesymnmf_fits_by_col[[i]] &lt;- codesymnmf::codesymnmf(cov_mat, 8)$H
}</code></pre>
<p>This is a heatmap of the estimated loadings from the first
subset:</p>
<pre class="r"><code>plot_heatmap(codesymnmf_fits_by_col[[1]], colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(codesymnmf_fits_by_col[[1]])), max(abs(codesymnmf_fits_by_col[[1]])), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-38-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-38-1">
Past versions of unnamed-chunk-38-1.png
</button>
</p>
<div id="fig-unnamed-chunk-38-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/stability_selection/blob/3532ad3890e7c167ffbaacc92157bcf5f97980a2/docs/figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-38-1.png" target="_blank">3532ad3</a>
</td>
<td>
Annie Xie
</td>
<td>
2025-09-11
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This is a heatmap of the estimated loadings from the second
subset:</p>
<pre class="r"><code>plot_heatmap(codesymnmf_fits_by_col[[2]], colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(codesymnmf_fits_by_col[[2]])), max(abs(codesymnmf_fits_by_col[[2]])), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-39-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-39-1">
Past versions of unnamed-chunk-39-1.png
</button>
</p>
<div id="fig-unnamed-chunk-39-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/xxie6/stability_selection/blob/3532ad3890e7c167ffbaacc92157bcf5f97980a2/docs/figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-39-1.png" target="_blank">3532ad3</a>
</td>
<td>
Annie Xie
</td>
<td>
2025-09-11
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>results_by_col &lt;- stability_selection_post_processing(codesymnmf_fits_by_col[[1]], codesymnmf_fits_by_col[[2]], threshold = 0.8)
L_est_by_col &lt;- results_by_col$L</code></pre>
<p>This is the similarity matrix:</p>
<pre class="r"><code>results_by_col$similarity</code></pre>
<pre><code>            [,1]        [,2]        [,3]        [,4]       [,5]       [,6]
[1,] 0.997942940 0.002771402 0.046514564 0.004185936 0.24057463 0.37769881
[2,] 0.008231438 0.998709784 0.023791189 0.019749744 0.78811821 0.02180517
[3,] 0.050261678 0.020372451 0.996653984 0.037347920 0.01419745 0.64407506
[4,] 0.018327960 0.016733425 0.031786751 0.998384167 0.20560138 0.05015960
[5,] 0.782662553 0.256198664 0.006487079 0.096357919 0.34486815 0.31574173
[6,] 0.041959889 0.548641605 0.592399354 0.028702038 0.42711935 0.33511395
[7,] 0.892769189 0.004787035 0.010234331 0.344844600 0.27324364 0.33722894
[8,] 0.517772790 0.012682400 0.621669476 0.025338730 0.09834010 0.62891987
            [,7]       [,8]
[1,] 0.433698519 0.71889091
[2,] 0.007712651 0.15045965
[3,] 0.065990718 0.25818496
[4,] 0.540919482 0.01199886
[5,] 0.329422374 0.51856580
[6,] 0.037082075 0.19937678
[7,] 0.561345163 0.64038338
[8,] 0.250712763 0.57207984</code></pre>
<pre class="r"><code>plot_heatmap(L_est_by_col, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L_est_by_col)), max(abs(L_est_by_col)), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-42-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This is the correlation between the estimate and the true loadings
matrix:</p>
<pre class="r"><code>apply(cor(L_est_by_col, sim_data$LL), 1, max)</code></pre>
<pre><code>[1] 0.9978969 0.9993604 0.9979943 0.9993792 0.8565386</code></pre>
<p>For a threshold of 0.8, the method recovers all four components.
However, it does return one extra factor.</p>
</div>
<div id="stability-selection-via-splitting-rows-2"
class="section level2">
<h2>Stability Selection via Splitting Rows</h2>
<pre class="r"><code>PolarU &lt;- function(A) {
  svdA &lt;- svd(A)
  out &lt;- svdA$u %*% t(svdA$v)
  return(out)
}</code></pre>
<pre class="r"><code>codesymnmf_fits_by_row_L &lt;- list()
codesymnmf_fits_by_row_F &lt;- list()
for (i in 1:(length(X_split_by_row)-2)){
  cov_mat &lt;- tcrossprod(X_split_by_row[[i]])/ncol(X_split_by_row[[i]])
  codesymnmf_fits_by_row_L[[i]] &lt;- codesymnmf::codesymnmf(cov_mat, 8)$H
  codesymnmf_fits_by_row_F[[i]] &lt;- PolarU(t(X_split_by_row[[i]]) %*% codesymnmf_fits_by_row_L[[i]])
}</code></pre>
<p>This is a heatmap of the estimated loadings from the first
subset:</p>
<pre class="r"><code>plot_heatmap(codesymnmf_fits_by_row_L[[1]][order(X_split_by_row[[3]]),], colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(codesymnmf_fits_by_row_L[[1]])), max(abs(codesymnmf_fits_by_row_L[[1]])), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-46-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This is a heatmap of the estimated loadings from the second
subset:</p>
<pre class="r"><code>plot_heatmap(codesymnmf_fits_by_row_L[[2]], colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(codesymnmf_fits_by_row_L[[2]])), max(abs(codesymnmf_fits_by_row_L[[2]])), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-47-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>results_by_row &lt;- stability_selection_row_split_post_processing(codesymnmf_fits_by_row_L[[1]], codesymnmf_fits_by_row_L[[2]], codesymnmf_fits_by_row_F[[1]], codesymnmf_fits_by_row_F[[2]], threshold = 0.6)
L_est_by_row &lt;- results_by_row$L
L_est_by_row &lt;- L_est_by_row[order(c(X_split_by_row[[3]], X_split_by_row[[4]])), ]</code></pre>
<p>This is the similarity matrix:</p>
<pre class="r"><code>results_by_row$similarity</code></pre>
<pre><code>              [,1]         [,2]          [,3]         [,4]         [,5]
[1,] -0.0051700996  0.959028407 -1.203145e-02 -0.007373688  0.037730172
[2,]  0.4650554717 -0.011296479  6.433738e-05 -0.006278056  0.014757182
[3,]  0.8578199706  0.002693785 -2.013129e-02  0.001903083  0.133793893
[4,] -0.0017412047 -0.006622036  9.710217e-01 -0.011926822  0.046969523
[5,] -0.0052247015 -0.011227756 -7.242105e-04  0.983479996 -0.001682197
[6,]  0.0003843685  0.082889010  1.164177e-01  0.029001273  0.008942243
[7,]  0.0053601553  0.182589365 -1.704194e-03  0.069593213 -0.007071516
[8,]  0.0762207432  0.007319490  9.739306e-02  0.038680721  0.017997563
            [,6]         [,7]         [,8]
[1,]  0.06214779  0.086907607  0.003300492
[2,]  0.07458203  0.030424592  0.035200771
[3,] -0.02913184 -0.003573936  0.046104545
[4,]  0.02803355  0.060528226  0.071199001
[5,]  0.02770146  0.011035023  0.051275416
[6,]  0.03212130 -0.019074611  0.016983648
[7,]  0.06221266 -0.004105695 -0.019475884
[8,]  0.02527426 -0.024892587  0.026923733</code></pre>
<p>This is a heatmap of the final loadings estimate:</p>
<pre class="r"><code>plot_heatmap(L_est_by_row, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L_est_by_row)), max(abs(L_est_by_row)), length.out = 50))</code></pre>
<p><img src="figure/bal_nonoverlap_setting.Rmd/unnamed-chunk-50-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This is the correlation between the estimate and the true loadings
matrix:</p>
<pre class="r"><code>apply(cor(L_est_by_row, sim_data$LL), 1, max)</code></pre>
<pre><code>[1] 0.9994426 0.9955690 0.9994670 0.9995463</code></pre>
<p>The method recovered all four components without returning any extra
factors.</p>
</div>
</div>
<div id="observations" class="section level1">
<h1>Observations</h1>
<p>Overall, all of the methods performed well in this setting. However,
there were a few instances of methods returning extra factors. This
analysis brought up the question of how to deal with redundant factors.
This is something I need to think about more. Whether extra factors are
returned also depends on the similarity threshold. I need to give more
consideration to how a similarity threshold is chosen.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS 15.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] pheatmap_1.0.12 ggplot2_3.5.2   dplyr_1.1.4     workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.1      viridisLite_0.4.2     farver_2.1.2         
 [4] fastmap_1.2.0         lazyeval_0.2.2        codesymnmf_0.0.0.9000
 [7] promises_1.3.3        digest_0.6.37         lifecycle_1.0.4      
[10] processx_3.8.4        invgamma_1.1          magrittr_2.0.3       
[13] compiler_4.3.2        rlang_1.1.6           sass_0.4.10          
[16] progress_1.2.3        tools_4.3.2           yaml_2.3.10          
[19] data.table_1.17.6     knitr_1.50            prettyunits_1.2.0    
[22] htmlwidgets_1.6.4     scatterplot3d_0.3-44  RColorBrewer_1.1-3   
[25] Rtsne_0.17            withr_3.0.2           purrr_1.0.4          
[28] flashier_1.0.56       grid_4.3.2            git2r_0.33.0         
[31] fastTopics_0.6-192    colorspace_2.1-1      scales_1.4.0         
[34] gtools_3.9.5          cli_3.6.5             rmarkdown_2.29       
[37] crayon_1.5.3          generics_0.1.4        RcppParallel_5.1.10  
[40] rstudioapi_0.16.0     httr_1.4.7            pbapply_1.7-2        
[43] cachem_1.1.0          stringr_1.5.1         splines_4.3.2        
[46] parallel_4.3.2        softImpute_1.4-3      vctrs_0.6.5          
[49] Matrix_1.6-5          jsonlite_2.0.0        callr_3.7.6          
[52] hms_1.1.3             mixsqp_0.3-54         ggrepel_0.9.6        
[55] irlba_2.3.5.1         horseshoe_0.2.0       trust_0.1-8          
[58] plotly_4.11.0         jquerylib_0.1.4       tidyr_1.3.1          
[61] ebcd_0.0.0.9000       glue_1.8.0            ebnm_1.1-34          
[64] ps_1.7.7              cowplot_1.1.3         gbcd_0.2-17          
[67] uwot_0.2.3            stringi_1.8.7         Polychrome_1.5.1     
[70] gtable_0.3.6          later_1.4.2           quadprog_1.5-8       
[73] tibble_3.3.0          pillar_1.10.2         htmltools_0.5.8.1    
[76] truncnorm_1.0-9       R6_2.6.1              rprojroot_2.0.4      
[79] evaluate_1.0.4        lattice_0.22-6        RhpcBLASctl_0.23-42  
[82] SQUAREM_2021.1        ashr_2.2-66           httpuv_1.6.15        
[85] bslib_0.9.0           Rcpp_1.0.14           deconvolveR_1.2-1    
[88] whisker_0.4.1         xfun_0.52             fs_1.6.6             
[91] getPass_0.2-4         pkgconfig_2.0.3      </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
