<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Annie Xie" />

<meta name="date" content="2025-09-11" />

<title>sparse_overlap_setting</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">stability_selection</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/xxie6/stability_selection">
    <span class="fab fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">sparse_overlap_setting</h1>
<h4 class="author">Annie Xie</h4>
<h4 class="date">2025-09-11</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-09-12
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>stability_selection/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20250906code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20250906)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20250906code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20250906)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomxxie6stabilityselectiontree017d47bc908c49c91687c06d51926976a46b966dtargetblank017d47ba">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/xxie6/stability_selection/tree/017d47bc908c49c91687c06d51926976a46b966d" target="_blank">017d47b</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomxxie6stabilityselectiontree017d47bc908c49c91687c06d51926976a46b966dtargetblank017d47ba"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/xxie6/stability_selection/tree/017d47bc908c49c91687c06d51926976a46b966d" target="_blank">017d47b</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/sparse_overlap_setting.Rmd</code>) and HTML
(<code>docs/sparse_overlap_setting.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/xxie6/stability_selection/blob/017d47bc908c49c91687c06d51926976a46b966d/analysis/sparse_overlap_setting.Rmd" target="_blank">017d47b</a>
</td>
<td>
Annie Xie
</td>
<td>
2025-09-12
</td>
<td>
Add exploration of sparse overlap setting
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>In this analysis, we are interested in testing stability selection
approaches in the sparse, overlapping setting.</p>
<p>At a high level, the stability selection involves 1) splitting the
data into two subsets, 2) applying the method to each subset, 3)
choosing the components that have high correspondence across the two
sets of results. We will test two different approaches to step 1). The
first approach is splitting the data by splitting the columns. This
approach feels intuitive since we are interested in the loadings matrix,
which says something about the samples in the dataset. In a population
genetics application, one could argue that all of the chromosomes are
undergoing evolution independently, and so you could split the data by
even vs. odd chromosomes to get two different datasets. However, in a
single-cell RNA-seq application, it feels more natural to split the data
by cells – this feels more like creating sub-datasets compared to
splitting by genes (unless you want to make some assumption that the
genes are pulled from a “population”, but I think that feels less
natural). This motivates the second approach: splitting the data by
splitting the rows.</p>
<p>I imagine that splitting the columns will be better than splitting
the rows in this setting because each sample is only active in a couple
of factors. Therefore, splitting by samples may lead some factors to
have weaker signal and thus be harder to find.</p>
<pre class="r"><code>library(dplyr)
library(ggplot2)
library(pheatmap)</code></pre>
<pre class="r"><code>source(&#39;code/visualization_functions.R&#39;)
source(&#39;code/stability_selection_functions.R&#39;)</code></pre>
<pre class="r"><code>permute_L &lt;- function(est, truth){
  K_est &lt;- ncol(est)
  K_truth &lt;- ncol(truth)
  n &lt;- nrow(est)

  #if estimates don&#39;t have same number of columns, try padding the estimate with zeros and make cosine similarity zero
  if (K_est &lt; K_truth){
    est &lt;- cbind(est, matrix(rep(0, n*(K_truth-K_est)), nrow = n))
  }

  if (K_est &gt; K_truth){
    truth &lt;- cbind(truth, matrix(rep(0, n*(K_est - K_truth)), nrow = n))
  }

  #normalize est and truth
  norms_est &lt;- apply(est, 2, function(x){sqrt(sum(x^2))})
  norms_est[norms_est == 0] &lt;- Inf

  norms_truth &lt;- apply(truth, 2, function(x){sqrt(sum(x^2))})
  norms_truth[norms_truth == 0] &lt;- Inf

  est_normalized &lt;- t(t(est)/norms_est)
  truth_normalized &lt;- t(t(truth)/norms_truth)

  #compute matrix of cosine similarities
  cosine_sim_matrix &lt;- abs(crossprod(est_normalized, truth_normalized))
  assignment_problem &lt;- lpSolve::lp.assign(t(cosine_sim_matrix), direction = &quot;max&quot;)

  perm &lt;- apply(assignment_problem$solution, 1, which.max)
  return(est[,perm])
}</code></pre>
</div>
<div id="data-generation" class="section level1">
<h1>Data Generation</h1>
<p>In this analysis, we will focus on the sparse overlapping
setting.</p>
<pre class="r"><code>sim_binary_loadings_data &lt;- function(args) {
  set.seed(args$seed)
  
  FF &lt;- matrix(rnorm(args$k * args$p, sd = args$group_sd), ncol = args$k)
  if (args$constrain_F) {
    FF_svd &lt;- svd(FF)
    FF &lt;- FF_svd$u
    FF &lt;- t(t(FF) * rep(args$group_sd, args$k) * sqrt(p))
  }
  LL &lt;- matrix(rbinom(args$n*args$k, 1, args$pi1), nrow = args$n, ncol = args$k)
  E &lt;- matrix(rnorm(args$n * args$p, sd = args$indiv_sd), nrow = args$n)
  
  Y &lt;- LL %*% t(FF) + E
  YYt &lt;- (1/args$p)*tcrossprod(Y)
  
  return(list(Y = Y, YYt = YYt, LL = LL, FF = FF, K = ncol(LL)))
}</code></pre>
<pre class="r"><code>n &lt;- 100
p &lt;- 1000
k &lt;- 10
pi1 &lt;- 0.1
indiv_sd &lt;- 1
group_sd &lt;- 1
seed &lt;- 1
sim_args = list(n = n, p = p, k = k, pi1 = pi1, indiv_sd = indiv_sd, group_sd = group_sd, seed = seed, constrain_F = FALSE)
sim_data &lt;- sim_binary_loadings_data(sim_args)</code></pre>
<p>This is a heatmap of the true loadings matrix:</p>
<pre class="r"><code>plot_heatmap(sim_data$LL)</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="gbcd" class="section level1">
<h1>GBCD</h1>
<p>In this section, I try stability selection with the GBCD method. In
my experiments, I’ve found that GBCD tends to return extra factors
(partially because the point-Laplace initialization will yield extra
factors).</p>
<div id="stability-selection-via-splitting-columns"
class="section level2">
<h2>Stability Selection via Splitting Columns</h2>
<p>First, I try splitting the data by splitting the columns.</p>
<pre class="r"><code>set.seed(1)
X_split_by_col &lt;- stability_selection_split_data(sim_data$Y, dim = &#39;columns&#39;)</code></pre>
<pre class="r"><code>gbcd_fits_by_col &lt;- list()
for (i in 1:length(X_split_by_col)){
  gbcd_fits_by_col[[i]] &lt;- gbcd::fit_gbcd(X_split_by_col[[i]], 
                                   Kmax = 10, 
                                   prior = ebnm::ebnm_generalized_binary,
                                   verbose = 0)$L
}</code></pre>
<pre><code>[1] &quot;Form cell by cell covariance matrix...&quot;
   user  system elapsed 
  0.004   0.000   0.004 
[1] &quot;Initialize GEP membership matrix L...&quot;
Adding factor 1 to flash object...
Wrapping up...
Done.
Adding factor 2 to flash object...
Adding factor 3 to flash object...
Adding factor 4 to flash object...
Adding factor 5 to flash object...
Adding factor 6 to flash object...
Adding factor 7 to flash object...
Adding factor 8 to flash object...
Adding factor 9 to flash object...
Adding factor 10 to flash object...
Wrapping up...
Done.</code></pre>
<pre><code>Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.</code></pre>
<pre><code>   user  system elapsed 
  4.068   0.133   4.210 
[1] &quot;Estimate GEP membership matrix L...&quot;
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 20 is numerically zero!
  --Estimate of factor 9 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 9 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 9 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 9 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 9 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 9 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 9 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 9 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 9 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 6 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 10 is numerically zero!</code></pre>
<pre><code>Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.</code></pre>
<pre><code>  --Estimate of factor 8 is numerically zero!
  --Estimate of factor 8 is numerically zero!
  --Estimate of factor 2 is numerically zero!
   user  system elapsed 
 13.633   0.323  14.071 
[1] &quot;Estimate GEP signature matrix F...&quot;
Backfitting 16 factors (tolerance: 7.45e-04)...
  --Estimate of factor 13 is numerically zero!
  --Estimate of factor 14 is numerically zero!
  --Estimate of factor 12 is numerically zero!
An update to factor 12 decreased the objective by 5.888e-04.
  Difference between iterations is within 1.0e+01...
  --Estimate of factor 12 is numerically zero!
An update to factor 12 decreased the objective by 4.755e-04.
An update to factor 12 decreased the objective by 2.141e-04.
  --Estimate of factor 12 is numerically zero!
An update to factor 12 decreased the objective by 2.006e-04.
  --Estimate of factor 12 is numerically zero!
An update to factor 12 decreased the objective by 4.800e-05.
  --Estimate of factor 12 is numerically zero!
  --Estimate of factor 9 is numerically zero!
  Difference between iterations is within 1.0e+00...
  --Maximum number of iterations reached!
Wrapping up...
Done.
   user  system elapsed 
 18.873   0.443  19.663 
[1] &quot;Form cell by cell covariance matrix...&quot;
   user  system elapsed 
  0.001   0.000   0.001 
[1] &quot;Initialize GEP membership matrix L...&quot;
Adding factor 1 to flash object...
Wrapping up...
Done.
Adding factor 2 to flash object...
Adding factor 3 to flash object...
Adding factor 4 to flash object...
Adding factor 5 to flash object...
Adding factor 6 to flash object...
Adding factor 7 to flash object...
Adding factor 8 to flash object...
Adding factor 9 to flash object...
Adding factor 10 to flash object...
Wrapping up...
Done.</code></pre>
<pre><code>Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.</code></pre>
<pre><code>   user  system elapsed 
  3.503   0.131   3.649 
[1] &quot;Estimate GEP membership matrix L...&quot;
  --Estimate of factor 17 is numerically zero!
  --Estimate of factor 20 is numerically zero!
  --Estimate of factor 10 is numerically zero!
  --Estimate of factor 14 is numerically zero!
  --Estimate of factor 9 is numerically zero!
  --Estimate of factor 16 is numerically zero!
  --Estimate of factor 6 is numerically zero!
  --Estimate of factor 2 is numerically zero!</code></pre>
<pre><code>Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.</code></pre>
<pre><code>   user  system elapsed 
  8.842   0.182   9.111 
[1] &quot;Estimate GEP signature matrix F...&quot;
Backfitting 11 factors (tolerance: 7.45e-04)...
  Difference between iterations is within 1.0e+01...
  --Estimate of factor 11 is numerically zero!
  Difference between iterations is within 1.0e+00...
  Difference between iterations is within 1.0e-01...
Wrapping up...
Done.
   user  system elapsed 
  0.831   0.021   0.852 </code></pre>
<p>This is heatmap of the loadings estimate from the first subset:</p>
<pre class="r"><code>L1_permuted &lt;- permute_L(gbcd_fits_by_col[[1]], sim_data$LL)
plot_heatmap(L1_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L1_permuted)), max(abs(L1_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This is a heatmap of the loadings estimate from the second
subset:</p>
<pre class="r"><code>L2_permuted &lt;- permute_L(gbcd_fits_by_col[[2]], sim_data$LL)
plot_heatmap(L2_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L2_permuted)), max(abs(L2_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>results_by_col &lt;- stability_selection_post_processing(gbcd_fits_by_col[[1]], gbcd_fits_by_col[[2]], threshold = 0.99)
L_est_by_col &lt;- results_by_col$L</code></pre>
<p>This is a heatmap of the final loadings estimate:</p>
<pre class="r"><code>L_est_by_col_permuted &lt;- permute_L(L_est_by_col, sim_data$LL)
L_est_by_col_permuted &lt;- L_est_by_col_permuted[, apply(L_est_by_col_permuted, 2, function(x){sum(x^2)}) &gt; 10^(-10)]
plot_heatmap(L_est_by_col_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L_est_by_col_permuted)), max(abs(L_est_by_col_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>In this case, the method recovered seven of the ten components.</p>
</div>
<div id="stability-selection-via-splitting-rows" class="section level2">
<h2>Stability Selection via Splitting Rows</h2>
<p>Now, we try splitting the rows:</p>
<pre class="r"><code>set.seed(1)
X_split_by_row &lt;- stability_selection_split_data(sim_data$Y, dim = &#39;rows&#39;)</code></pre>
<pre class="r"><code>gbcd_fits_by_row &lt;- list()
for (i in 1:length(X_split_by_row)){
  gbcd_fits_by_row[[i]] &lt;- gbcd::fit_gbcd(X_split_by_row[[i]], 
                                   Kmax = 10, 
                                   prior = ebnm::ebnm_generalized_binary,
                                   verbose = 0)$L
}</code></pre>
<pre><code>[1] &quot;Form cell by cell covariance matrix...&quot;
   user  system elapsed 
  0.000   0.000   0.001 
[1] &quot;Initialize GEP membership matrix L...&quot;
Adding factor 1 to flash object...
Wrapping up...
Done.
Adding factor 2 to flash object...
Adding factor 3 to flash object...
Adding factor 4 to flash object...
Adding factor 5 to flash object...
Adding factor 6 to flash object...
Adding factor 7 to flash object...
Adding factor 8 to flash object...
Adding factor 9 to flash object...
Adding factor 10 to flash object...
Wrapping up...
Done.</code></pre>
<pre><code>Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.</code></pre>
<pre><code>   user  system elapsed 
  5.785   0.190   5.974 
[1] &quot;Estimate GEP membership matrix L...&quot;</code></pre>
<pre><code>Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.</code></pre>
<pre><code>   user  system elapsed 
 35.470   0.957  37.058 
[1] &quot;Estimate GEP signature matrix F...&quot;
Backfitting 16 factors (tolerance: 7.45e-05)...
  Difference between iterations is within 1.0e+00...
  Difference between iterations is within 1.0e-01...
  Difference between iterations is within 1.0e-02...
  Difference between iterations is within 1.0e-03...
  Difference between iterations is within 1.0e-04...
Wrapping up...
Done.
   user  system elapsed 
  0.538   0.009   0.547 
[1] &quot;Form cell by cell covariance matrix...&quot;
   user  system elapsed 
  0.003   0.000   0.002 
[1] &quot;Initialize GEP membership matrix L...&quot;
Adding factor 1 to flash object...
Wrapping up...
Done.
Adding factor 2 to flash object...
Adding factor 3 to flash object...
Adding factor 4 to flash object...
Adding factor 5 to flash object...
Adding factor 6 to flash object...
Adding factor 7 to flash object...
Adding factor 8 to flash object...
Adding factor 9 to flash object...
Adding factor 10 to flash object...
Wrapping up...
Done.</code></pre>
<pre><code>Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.</code></pre>
<pre><code>   user  system elapsed 
  2.743   0.095   2.842 
[1] &quot;Estimate GEP membership matrix L...&quot;
  --Estimate of factor 18 is numerically zero!
  --Estimate of factor 19 is numerically zero!
  --Estimate of factor 8 is numerically zero!
  --Estimate of factor 2 is numerically zero!
  --Estimate of factor 6 is numerically zero!
  --Estimate of factor 2 is numerically zero!
  --Estimate of factor 2 is numerically zero!
  --Estimate of factor 2 is numerically zero!
  --Estimate of factor 2 is numerically zero!</code></pre>
<pre><code>Warning in report.maxiter.reached(verbose.lvl): Maximum number of iterations
reached.</code></pre>
<pre><code>  --Estimate of factor 5 is numerically zero!
   user  system elapsed 
 13.273   0.278  13.761 
[1] &quot;Estimate GEP signature matrix F...&quot;
Backfitting 12 factors (tolerance: 1.42e-03)...
  Difference between iterations is within 1.0e+02...
  --Estimate of factor 12 is numerically zero!
  --Estimate of factor 4 is numerically zero!
  Difference between iterations is within 1.0e+01...
  Difference between iterations is within 1.0e+00...
  Difference between iterations is within 1.0e-01...
Wrapping up...
Done.
   user  system elapsed 
  1.665   0.038   1.705 </code></pre>
<p>This is heatmap of the loadings estimate from the first subset:</p>
<pre class="r"><code>L1_permuted &lt;- permute_L(gbcd_fits_by_row[[1]], sim_data$LL)
plot_heatmap(L1_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L1_permuted)), max(abs(L1_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-15-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This is heatmap of the loadings estimate from the second subset:</p>
<pre class="r"><code>L2_permuted &lt;- permute_L(gbcd_fits_by_row[[2]], sim_data$LL)
plot_heatmap(L2_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L2_permuted)), max(abs(L2_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-16-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>results_by_row &lt;- stability_selection_post_processing(gbcd_fits_by_row[[1]], gbcd_fits_by_row[[2]], threshold = 0.99)
L_est_by_row &lt;- results_by_row$L</code></pre>
<p>This is a heatmap of the final loadings estimate:</p>
<pre class="r"><code>L_est_by_row_permuted &lt;- permute_L(L_est_by_row, sim_data$LL)
L_est_by_row_permuted &lt;- L_est_by_row_permuted[, apply(L_est_by_row_permuted, 2, function(x){sum(x^2)}) &gt; 10^(-10)]
plot_heatmap(L_est_by_row_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L_est_by_row_permuted)), max(abs(L_est_by_row_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-18-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>In this case, the method recovered two of the ten factors.</p>
</div>
</div>
<div id="ebcd" class="section level1">
<h1>EBCD</h1>
<p>In this section, I try stability selection with the GBCD method. When
given a <code>Kmax</code> value that is larger than the true number of
components, I’ve found that EBCD usually returns extra factors. So in
this section, when I run EBCD, I give the method double the true number
of components.</p>
<div id="stability-selection-via-splitting-columns-1"
class="section level2">
<h2>Stability Selection via Splitting Columns</h2>
<pre class="r"><code>set.seed(1)
ebcd_fits_by_col &lt;- list()
for (i in 1:length(X_split_by_col)){
  ebcd_fits_by_col[[i]] &lt;- ebcd::ebcd(t(X_split_by_col[[i]]), 
                                   Kmax = 20, 
                                   ebnm_fn = ebnm::ebnm_generalized_binary)$EL
}</code></pre>
<p>This is heatmap of the loadings estimate from the first subset:</p>
<pre class="r"><code>L1_permuted &lt;- permute_L(ebcd_fits_by_col[[1]], sim_data$LL)
plot_heatmap(L1_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L1_permuted)), max(abs(L1_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This is a heatmap of the loadings estimate from the second
subset:</p>
<pre class="r"><code>L2_permuted &lt;- permute_L(ebcd_fits_by_col[[2]], sim_data$LL)
plot_heatmap(L2_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L2_permuted)), max(abs(L2_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>results_by_col &lt;- stability_selection_post_processing(ebcd_fits_by_col[[1]], ebcd_fits_by_col[[2]], threshold = 0.99)
L_est_by_col &lt;- results_by_col$L</code></pre>
<p>This is a heatmap of the final loadings estimate:</p>
<pre class="r"><code>L_est_by_col_permuted &lt;- permute_L(L_est_by_col, sim_data$LL)
L_est_by_col_permuted &lt;- L_est_by_col_permuted[, apply(L_est_by_col_permuted, 2, function(x){sum(x^2)}) &gt; 10^(-10)]
plot_heatmap(L_est_by_col_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L_est_by_col_permuted)), max(abs(L_est_by_col_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-23-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>These are the correlations between the estimates and the true factors
(which are paired to maximize crossproduct similarity):</p>
<pre class="r"><code>diag(cor(L_est_by_col_permuted, sim_data$LL))</code></pre>
<pre><code> [1] 0.9975203 0.9974854 0.9983737 0.9988182 0.9992998 0.9991493 0.9990425
 [8] 0.9982839 0.9981471 0.9978499</code></pre>
<p>The method recovers all ten components nearly perfectly.</p>
</div>
<div id="stability-selection-via-splitting-rows-1"
class="section level2">
<h2>Stability Selection via Splitting Rows</h2>
<pre class="r"><code>set.seed(1)
ebcd_fits_by_row &lt;- list()
for (i in 1:length(X_split_by_row)){
  ebcd_fits_by_row[[i]] &lt;- ebcd::ebcd(t(X_split_by_row[[i]]), 
                                   Kmax = 20, 
                                   ebnm_fn = ebnm::ebnm_generalized_binary)$EL
}</code></pre>
<p>This is heatmap of the loadings estimate from the first subset:</p>
<pre class="r"><code>L1_permuted &lt;- permute_L(ebcd_fits_by_row[[1]], sim_data$LL)
plot_heatmap(L1_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L1_permuted)), max(abs(L1_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-26-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This is heatmap of the loadings estimate from the second subset:</p>
<pre class="r"><code>L2_permuted &lt;- permute_L(ebcd_fits_by_row[[2]], sim_data$LL)
plot_heatmap(L2_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L2_permuted)), max(abs(L2_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-27-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>results_by_row &lt;- stability_selection_post_processing(ebcd_fits_by_row[[1]], ebcd_fits_by_row[[2]], threshold = 0.99)
L_est_by_row &lt;- results_by_row$L</code></pre>
<p>This is a heatmap of the final loadings estimate:</p>
<pre class="r"><code>L_est_by_row_permuted &lt;- permute_L(L_est_by_row, sim_data$LL)
L_est_by_row_permuted &lt;- L_est_by_row_permuted[, apply(L_est_by_row_permuted, 2, function(x){sum(x^2)}) &gt; 10^(-10)]
plot_heatmap(L_est_by_row_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L_est_by_row_permuted)), max(abs(L_est_by_row_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-29-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>In this case, the method returns only seven of the ten factors.</p>
</div>
</div>
<div id="codesymnmf" class="section level1">
<h1>CoDesymNMF</h1>
<p>In this section, I try stability selection with the CoDesymNMF
method. Similar to EBCD, when given a <code>Kmax</code> value that is
larger than the true number of components, the method usually returns
extra factors. Note that in this section, when I run CoDesymNMF, I give
the method double the true number of components.</p>
<div id="stability-selection-via-splitting-columns-2"
class="section level2">
<h2>Stability Selection via Splitting Columns</h2>
<pre class="r"><code>codesymnmf_fits_by_col &lt;- list()
for (i in 1:length(X_split_by_col)){
  cov_mat &lt;- tcrossprod(X_split_by_col[[i]])/ncol(X_split_by_col[[i]])
  codesymnmf_fits_by_col[[i]] &lt;- codesymnmf::codesymnmf(cov_mat, 20)$H
}</code></pre>
<p>This is heatmap of the loadings estimate from the first subset:</p>
<pre class="r"><code>L1_permuted &lt;- permute_L(codesymnmf_fits_by_col[[1]], sim_data$LL)
plot_heatmap(L1_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L1_permuted)), max(abs(L1_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-31-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This is a heatmap of the loadings estimate from the second
subset:</p>
<pre class="r"><code>L2_permuted &lt;- permute_L(codesymnmf_fits_by_col[[2]], sim_data$LL)
plot_heatmap(L2_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L2_permuted)), max(abs(L2_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-32-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>results_by_col &lt;- stability_selection_post_processing(codesymnmf_fits_by_col[[1]], codesymnmf_fits_by_col[[2]], threshold = 0.99)
L_est_by_col &lt;- results_by_col$L</code></pre>
<p>In this case, the method does not return any factors. Reducing the
similarity threshold to 0.98 allows the method to return four of the ten
factors. So it seems like the estimates do contain the signal. However,
perhaps due to the fact the method doesn’t encourage sparsity or is not
binary, the estimates are not exactly the same across the two
estimates.</p>
<pre class="r"><code>results_by_col &lt;- stability_selection_post_processing(codesymnmf_fits_by_col[[1]], codesymnmf_fits_by_col[[2]], threshold = 0.98)
L_est_by_col &lt;- results_by_col$L</code></pre>
<p>This is a heatmap of the final loadings estimate with a decreased
threshold:</p>
<pre class="r"><code>L_est_by_col_permuted &lt;- permute_L(L_est_by_col, sim_data$LL)
L_est_by_col_permuted &lt;- L_est_by_col_permuted[, apply(L_est_by_col_permuted, 2, function(x){sum(x^2)}) &gt; 10^(-10)]
plot_heatmap(L_est_by_col_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L_est_by_col_permuted)), max(abs(L_est_by_col_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-35-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="stability-selection-via-splitting-rows-2"
class="section level2">
<h2>Stability Selection via Splitting Rows</h2>
<pre class="r"><code>codesymnmf_fits_by_row &lt;- list()
for (i in 1:length(X_split_by_row)){
  cov_mat &lt;- tcrossprod(X_split_by_row[[i]])/ncol(X_split_by_row[[i]])
  codesymnmf_fits_by_row[[i]] &lt;- codesymnmf::codesymnmf(cov_mat, 20)$H
}</code></pre>
<p>This is heatmap of the loadings estimate from the first subset:</p>
<pre class="r"><code>L1_permuted &lt;- permute_L(codesymnmf_fits_by_row[[1]], sim_data$LL)
plot_heatmap(L1_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L1_permuted)), max(abs(L1_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-37-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This is heatmap of the loadings estimate from the second subset:</p>
<pre class="r"><code>L2_permuted &lt;- permute_L(codesymnmf_fits_by_row[[2]], sim_data$LL)
plot_heatmap(L2_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L2_permuted)), max(abs(L2_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-38-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>results_by_row &lt;- stability_selection_post_processing(codesymnmf_fits_by_row[[1]], codesymnmf_fits_by_row[[2]], threshold = 0.99)
L_est_by_row &lt;- results_by_row$L</code></pre>
<p>In this setting, the method again does not return any factors. The
method only starts returning factors when the threshold is reduced to
0.92, so this suggests the estimates are more dissimilar in this
case.</p>
<pre class="r"><code>results_by_row &lt;- stability_selection_post_processing(codesymnmf_fits_by_row[[1]], codesymnmf_fits_by_row[[2]], threshold = 0.92)
L_est_by_row &lt;- results_by_row$L</code></pre>
<p>This is a heatmap of the final loadings estimate with a reduced
threshold:</p>
<pre class="r"><code>L_est_by_row_permuted &lt;- permute_L(L_est_by_row, sim_data$LL)
L_est_by_row_permuted &lt;- L_est_by_row_permuted[, apply(L_est_by_row_permuted, 2, function(x){sum(x^2)}) &gt; 10^(-10)]
plot_heatmap(L_est_by_row_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L_est_by_row_permuted)), max(abs(L_est_by_row_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-41-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="ebmfcov" class="section level1">
<h1>EBMFcov</h1>
<p>In this section, I try stability selection with the EBMFcov method.
In my experiments, I’ve found that when given a larger
<code>Kmax</code>, EBMFcov will get around the correct number of
factors, and then it will add essentially trivial factors. I wonder if
the stability selection procedure will help get rid of these extra
factors. The hope is that it can. However, if both estimates have
factors that look like this, then maybe these factors will still get
returned. Note that in this section, when I run EBMFcov, I give the
method double the true number of components.</p>
<pre class="r"><code>cov_fit &lt;- function(covmat, ebnm_fn = ebnm::ebnm_point_laplace, Kmax = 1000, verbose.lvl = 0, backfit = TRUE) {
  fl &lt;- flash_init(covmat, var_type = 0) %&gt;%
    flash_set_verbose(verbose.lvl) %&gt;%
    flash_greedy(ebnm_fn = ebnm_fn, Kmax = Kmax)
  if (backfit == TRUE){
    fl &lt;- flash_backfit(fl)
  }
  s2 &lt;- max(0, mean(diag(covmat) - diag(fitted(fl))))
  s2_diff &lt;- Inf
  while(s2 &gt; 0 &amp;&amp; abs(s2_diff - 1) &gt; 1e-4) {
    covmat_minuss2 &lt;- covmat - diag(rep(s2, ncol(covmat)))
    fl &lt;- flash_init(covmat_minuss2, var_type = 0) %&gt;%
      flash_set_verbose(verbose.lvl) %&gt;%
      flash_greedy(ebnm_fn = ebnm_fn, Kmax = Kmax)
    if (backfit == TRUE){
      fl &lt;- flash_backfit(fl)
    }
    old_s2 &lt;- s2
    s2 &lt;- max(0, mean(diag(covmat) - diag(fitted(fl))))
    s2_diff &lt;- s2 / old_s2
  }
  
  return(list(fl=fl, s2 = s2))
}</code></pre>
<div id="stability-selection-via-splitting-columns-3"
class="section level2">
<h2>Stability Selection via Splitting Columns</h2>
<pre class="r"><code>library(flashier)</code></pre>
<pre><code>Loading required package: ebnm</code></pre>
<pre class="r"><code>ebmfcov_fits_by_col &lt;- list()
for (i in 1:length(X_split_by_col)){
  cov_mat &lt;- tcrossprod(X_split_by_col[[i]])/ncol(X_split_by_col[[i]])
  fl &lt;- cov_fit(cov_mat, ebnm_fn = ebnm::ebnm_generalized_binary, Kmax = 20)$fl
  fl_scaled &lt;- ldf(fl)
  L_scaled &lt;- fl_scaled$L %*% diag(sqrt(fl_scaled$D))
  ebmfcov_fits_by_col[[i]] &lt;- L_scaled
}</code></pre>
<p>Just a note: this took a really long time to run. I didn’t expect it
to take so long.</p>
<p>This is heatmap of the loadings estimate from the first subset:</p>
<pre class="r"><code>L1_permuted &lt;- permute_L(ebmfcov_fits_by_col[[1]], sim_data$LL)
plot_heatmap(L1_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L1_permuted)), max(abs(L1_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-45-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This is a heatmap of the loadings estimate from the second
subset:</p>
<pre class="r"><code>L2_permuted &lt;- permute_L(ebmfcov_fits_by_col[[2]], sim_data$LL)
plot_heatmap(L2_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L2_permuted)), max(abs(L2_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-46-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>results_by_col &lt;- stability_selection_post_processing(ebmfcov_fits_by_col[[1]], ebmfcov_fits_by_col[[2]], threshold = 0.99)
L_est_by_col &lt;- results_by_col$L</code></pre>
<p>This is a heatmap of the final loadings estimate:</p>
<pre class="r"><code>L_est_by_col_permuted &lt;- permute_L(L_est_by_col, sim_data$LL)
L_est_by_col_permuted &lt;- L_est_by_col_permuted[, apply(L_est_by_col_permuted, 2, function(x){sum(x^2)}) &gt; 10^(-10)]
plot_heatmap(L_est_by_col_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L_est_by_col_permuted)), max(abs(L_est_by_col_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-48-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This is a heatmap of the factors in the final loadings estimate that
appear to be trivial:</p>
<pre class="r"><code>plot_heatmap(L_est_by_col_permuted[, c(3, 6, 11:19)], colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L_est_by_col_permuted[, c(3, 6, 11:19)])), max(abs(L_est_by_col_permuted[, c(3, 6, 11:19)])), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-49-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>So it appears that the method still returns since both estimates have
factors of this type. I think this is something specific to EBMFcov and
the generalized binary prior when <code>Kmax</code> is large, so perhaps
another type of ad-hoc procedure is needed to get rid of these extra
factors.</p>
</div>
<div id="stability-selection-via-splitting-rows-3"
class="section level2">
<h2>Stability Selection via Splitting Rows</h2>
<pre class="r"><code>ebmfcov_fits_by_row &lt;- list()
for (i in 1:length(X_split_by_row)){
  cov_mat &lt;- tcrossprod(X_split_by_row[[i]])/ncol(X_split_by_row[[i]])
  fl &lt;- cov_fit(cov_mat, ebnm_fn = ebnm::ebnm_generalized_binary, Kmax = 20)$fl
  fl_scaled &lt;- ldf(fl)
  L_scaled &lt;- fl_scaled$L %*% diag(sqrt(fl_scaled$D))
  ebmfcov_fits_by_row[[i]] &lt;- L_scaled
}</code></pre>
<p>This is heatmap of the loadings estimate from the first subset:</p>
<pre class="r"><code>L1_permuted &lt;- permute_L(ebmfcov_fits_by_row[[1]], sim_data$LL)
plot_heatmap(L1_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L1_permuted)), max(abs(L1_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-51-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This is heatmap of the loadings estimate from the second subset:</p>
<pre class="r"><code>L2_permuted &lt;- permute_L(ebmfcov_fits_by_row[[2]], sim_data$LL)
plot_heatmap(L2_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L2_permuted)), max(abs(L2_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-52-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>results_by_row &lt;- stability_selection_post_processing(ebmfcov_fits_by_row[[1]], ebmfcov_fits_by_row[[2]], threshold = 0.99)
L_est_by_row &lt;- results_by_row$L</code></pre>
<p>This is a heatmap of the final loadings estimate:</p>
<pre class="r"><code>L_est_by_row_permuted &lt;- permute_L(L_est_by_row, sim_data$LL)
L_est_by_row_permuted &lt;- L_est_by_row_permuted[, apply(L_est_by_row_permuted, 2, function(x){sum(x^2)}) &gt; 10^(-10)]
plot_heatmap(L_est_by_row_permuted, colors_range = c(&#39;blue&#39;,&#39;gray96&#39;,&#39;red&#39;), brks = seq(-max(abs(L_est_by_row_permuted)), max(abs(L_est_by_row_permuted)), length.out = 50))</code></pre>
<p><img src="figure/sparse_overlap_setting.Rmd/unnamed-chunk-54-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The method returns three of the ten factors. Again, the method also
returns some of the extra factors.</p>
</div>
</div>
<div id="observations" class="section level1">
<h1>Observations</h1>
<p>As expected, the variant of stability selection which split the data
via the rows struggled more than the variant which split the data via
the columns. I thought it was interesting that EBCD with the
column-splitting based stability selection was able to return all ten
factors – this suggests that when given a higher <code>Kmax</code>, EBCD
will return extra factors on top of the correct factors (as opposed to
representing the matrix with a characteristically different
representation that utilizes its full factor allotment). CoDesymNMF did
not return any factors with either stability selection method. This may
be a result of the lack of a sparsity penalty in CoDesymNMF.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS 15.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] flashier_1.0.56 ebnm_1.1-34     pheatmap_1.0.12 ggplot2_3.5.2  
[5] dplyr_1.1.4     workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.1      viridisLite_0.4.2     farver_2.1.2         
 [4] fastmap_1.2.0         lazyeval_0.2.2        codesymnmf_0.0.0.9000
 [7] promises_1.3.3        digest_0.6.37         lifecycle_1.0.4      
[10] processx_3.8.4        invgamma_1.1          magrittr_2.0.3       
[13] compiler_4.3.2        rlang_1.1.6           sass_0.4.10          
[16] progress_1.2.3        tools_4.3.2           yaml_2.3.10          
[19] data.table_1.17.6     knitr_1.50            prettyunits_1.2.0    
[22] htmlwidgets_1.6.4     scatterplot3d_0.3-44  RColorBrewer_1.1-3   
[25] Rtsne_0.17            withr_3.0.2           purrr_1.0.4          
[28] grid_4.3.2            git2r_0.33.0          fastTopics_0.6-192   
[31] colorspace_2.1-1      scales_1.4.0          gtools_3.9.5         
[34] cli_3.6.5             rmarkdown_2.29        crayon_1.5.3         
[37] generics_0.1.4        RcppParallel_5.1.10   rstudioapi_0.16.0    
[40] httr_1.4.7            pbapply_1.7-2         cachem_1.1.0         
[43] stringr_1.5.1         splines_4.3.2         parallel_4.3.2       
[46] softImpute_1.4-3      vctrs_0.6.5           Matrix_1.6-5         
[49] jsonlite_2.0.0        callr_3.7.6           hms_1.1.3            
[52] mixsqp_0.3-54         ggrepel_0.9.6         irlba_2.3.5.1        
[55] horseshoe_0.2.0       trust_0.1-8           plotly_4.11.0        
[58] jquerylib_0.1.4       tidyr_1.3.1           ebcd_0.0.0.9000      
[61] glue_1.8.0            ps_1.7.7              cowplot_1.1.3        
[64] gbcd_0.2-17           uwot_0.2.3            stringi_1.8.7        
[67] Polychrome_1.5.1      gtable_0.3.6          later_1.4.2          
[70] quadprog_1.5-8        tibble_3.3.0          pillar_1.10.2        
[73] htmltools_0.5.8.1     truncnorm_1.0-9       R6_2.6.1             
[76] rprojroot_2.0.4       lpSolve_5.6.20        evaluate_1.0.4       
[79] lattice_0.22-6        RhpcBLASctl_0.23-42   SQUAREM_2021.1       
[82] ashr_2.2-66           httpuv_1.6.15         bslib_0.9.0          
[85] Rcpp_1.0.14           deconvolveR_1.2-1     whisker_0.4.1        
[88] xfun_0.52             fs_1.6.6              getPass_0.2-4        
[91] pkgconfig_2.0.3      </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
